File: scala/parsers/src/test/scala/jp/co/teruuu/parser/common/OptionalParserTest.scala
package jp.co.teruuu.parser.common

import org.scalatest.FlatSpec;

class OptionalParserTest extends FlatSpec {

  "parse" should "successt" in {
    val parser = Parser.string("Hello").option()
    val result1 = parser.parse("Hello, World")
    assert(result1.isInstanceOf[ParseOk[_]])
    assert(result1.asInstanceOf[ParseOk[Option[String]]].value.equals(Some("Hello")))

    val result2 = parser.parse("Hell, World")
    assert(result2.isInstanceOf[ParseOk[_]])
    assert(result2.asInstanceOf[ParseOk[Option[String]]].value.equals(None))
  }

}

File: java/parsers_java8/src/test/java/jp/co/teruuu/parser/common/OptionalParserTest.java
package jp.co.teruuu.parser.common;

import jp.co.teruuu.parser.common.ParseResult;
import jp.co.teruuu.parser.common.Parser;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class OptionalParserTest {

    @Test
    public void test() {
        Parser<Optional<List<String>>> parser = Parser.string("Hello").plus().option();

        ParseResult result1 = parser.parse("HelloHelloHello");
        assert(result1 instanceof ParseResult.Success);
        assert(((ParseResult.Success<Optional<List<String>>>) result1).value.isPresent());
        assert(((ParseResult.Success<Optional<List<String>>>) result1).value.get().equals(Arrays.asList("Hello", "Hello", "Hello")));

        ParseResult result2 = parser.parse("Hello HelloHello");
        assert(result2 instanceof ParseResult.Success);
        assert(((ParseResult.Success<Optional<List<String>>>) result2).value.isPresent());
        assert(((ParseResult.Success<Optional<List<String>>>) result2).value.get().equals(Arrays.asList("Hello")));


        ParseResult result3 = parser.parse("helloHelloHello");
        assert(result3 instanceof ParseResult.Success);
        assert(!((ParseResult.Success<Optional<List<String>>>) result3).value.isPresent());
    }
}
